qiskit/pulse/instructions/instruction.py

```diff
@@ -59,20 +58,18 @@ def __init__(self,
             PulseError: If the input ``channels`` are not all of
                 type :class:`Channel`.
         """
-        if not isinstance(duration, (int, np.integer)):
-            raise PulseError("Instruction duration must be an integer, "
-                             "got {} instead.".format(duration))
-        if duration < 0:
-            raise PulseError("{} duration of {} is invalid: must be nonnegative."
-                             "".format(self.__class__.__name__, duration))
-
         for channel in channels:
             if not isinstance(channel, Channel):
                 raise PulseError("Expected a channel, got {} instead.".format(channel))
 
-        self._duration = duration
+        if duration is not None:
+            warnings.warn('Specifying duration in the constructor is deprecated. '
+                          'Now duration is abstract property rather than class variable. '
+                          'All subclasses should implement ``duration`` accordingly. '
+                          'See Qiskit-Terra #5679 for more information.',
+                          DeprecationWarning)
+
         self._channels = channels
-        self._timeslots = {channel: [(0, self.duration)] for channel in channels}
```

Comment:
Maybe `_timeslots` private attribute should remain, to make things better in `Schedule`?

Reply:
I think instructions are no longer schedule component, i.e. they don't have notion of `t0`, so having timeslots in this class seems to be contradictory to this concept. I agree with calling the `_get_timeslots` function is little bit of overhead in `Schedule`.

Reply:
Right, I agree. And it should be ok since Instructions only have a few channels 👍